[kostalinverter]  Extended with functionality for Second Generation (Piko10-20) type inverters.

basse04 opened the pull request:
Binding redesigned 20200924 by Örjan Backsell

Signed-off-by: basse04 orjan.backsell@gmail.com
Also-by: Christian Schneider <>
Also-by: René Stakemeier <>
Also-by: Christoph Weitkamp <>


This is the upgraded version of Kostal Inverter extended with functionality for the Inverter type of Second generation PIKO 10-20
[kostalinverter] regarded oH3.

This is the upgraded version of [kostalinverter] extended with functionality for Kostal  Piko 10-20 (Second Generation) inverters.
This PR is a replacement of #7796 which was closed due to the switch from 2.5.x to 3.0.

BR Basse

basse04 commented:
Done changes in org.openhab.binding.internal.kostal.inverter.secondgeneration.SecondGenerationHandler.java regarded to Travis tests.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+
+    public static final long REFRESHINTERVAL = 60;
+
+    public Object url = "";
```
Can this be String?

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+
+        if (command instanceof RefreshType) {
+            logger.debug("Handle command for {} on channel {}: {}", thing.getUID(), channelUID, command);
+        }
+
+        if (command == RefreshType.REFRESH) {
+            return;
+        }
```
These could be merged to a single `if`. Can you handle the REFRESH command? You could use `ExpiringCache` to reduce the invocations of `refresh()`, as REFRESH is triggered for each Channel.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+            try {
+                updateStatus(ThingStatus.ONLINE);
+                refresh();
+            } catch (RuntimeException scheduleWithFixedDelayException) {
```
What type of exception do you expect here? Can you specify the concrete type?

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+
+        scheduler.scheduleWithFixedDelay(() -> {
+            try {
+                updateStatus(ThingStatus.ONLINE);
```
Better set the Thing to ONLINE after the communication has succeeded. This avoids flickering of the state.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationDxsEntriesContainer.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Örjan Backsell - Initial contribution Piko1020, Piko New Generation
+ */
+
+@NonNullByDefault
```
DTOs don't need to be annotated. You could append `DTO` to the class name to avoid the checkstyle warning. Then, you can also remove `@SuppressWarnings` of `refresh()` in the handler.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+        try {
+            SecondGenerationConfigurationHandler.executeConfigurationChanges(httpClient, url, username, password,
+                    dxsEntriesConf, valueConfiguration);
+        } catch (Exception handleCommandException) {
```
You should specify the concrete exception you want to catch, since catching `Exception` catches also all runtime exceptions.

fwolter replied:
Can you comment on this?

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+        }
+    }
+
+    public final String callURL(String myURL) {
```
Is there a reason why you don't use the `httpClient`?

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,27 @@
+	xsi:schemaLocation="https://openhab.org/schemas/config-description/v1.0.0 https://openhab.org/schemas/config-description-1.0.0.xsd">
+	<config-description uri="thing-type:kostalpiko1020_config">
+		<parameter name="url" type="text" required="true">
+			<label> Inverter Url</label>
```
```suggestion
			<label>Inverter URL</label>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/PIKO1020.xml:
```diff
@@ -0,0 +1,81 @@
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<thing-type id="kostalinverterpiko1020">
```
The binding name don't need to be included in the Thing Type ID.
```suggestion
	<thing-type id="piko1020">
```

fwolter replied:
See my comment from 7 Oct. "The binding name don't need to be included in the Thing Type ID." Is this the parameter you meant, preventing the Thing from being showed up in the UI? Actually this naming shouldn't prevent it from showing up.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/PIKO1020.xml:
```diff
@@ -0,0 +1,81 @@
+		<description>Thing for the KOSTAL PIKO 10-20</description>
+		<category>Inverter</category>
+		<channels>
+			<channel typeId="device-local-grid-output-power" id="gridOutputPower"/>
```
Some of these Channel typeIds doesn't match with those in Channels.xml.

fwolter commented:
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

basse04 commented:
Thank's, done some of the instructed changes.

basse04 commented:
Done some more of instructed changes.

basse04 commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,354 @@
+                SecondGenerationDxsEntriesContainer.class);
+
+        // Create channel-posts array's
+        String[] channelPosts = new String[23];
+        String[] channelPostsExt = new String[23];
+        String[] channelPostsExtExt = new String[3];
+
+        // Fill channelPosts with each item value
+        int channelPostsCounter = 0;
+        for (SecondGenerationDxsEntries dxsentries : dxsEntriesContainer.dxsEntries) {
+            channelPosts[channelPostsCounter] = dxsentries.getName();
+            channelPostsCounter++;
+        }
+
+        // Fill channelPostsExt with each item value
+        int channelPostsCounterExt = 0;
+        for (SecondGenerationDxsEntries dxsentriesExt : dxsEntriesContainerExt.dxsEntries) {
+            channelPostsExt[channelPostsCounterExt] = dxsentriesExt.getName();
+            channelPostsCounterExt++;
+        }
+
+        // Fill channelPostsExtExt with each item value
+        int channelPostsCounterExtExt = 0;
+        for (SecondGenerationDxsEntries dxsentriesExtExt : dxsEntriesContainerExtExt.dxsEntries) {
+            channelPostsExtExt[channelPostsCounterExtExt] = dxsentriesExtExt.getName();
+            channelPostsCounterExtExt++;
+        }
```
It's only possible to scrape maximum 23 values (dxsId's)  each time from the inverter.

cpmeister replied:
I wasn't talking about scraping the values, I was referring to aggregating the retrieved values into a single array/list.
Something like this:
```suggestion
        List<String> channelPosts = new ArrayList<>();
        channelPosts.addAll(dxsEntriesContainer.dxsEntries.stream());
        channelPosts.addAll(dxsEntriesContainerExt.dxsEntries.stream());
        channelPosts.addAll(dxsEntriesContainerExtExt.dxsEntries.stream());
```

basse04 replied:
Sorry, for the misunderstanding. I will have a look and be back asap.

basse04 commented:
Done changes in org.openhab.binding.internal.kostal.inverter.secondgeneration.x, regarded to @fwolter requested changes.

fwolter commented:
Did you incorporate all review feedback? Asking because the conversations aren't marked as resolved.

basse04 commented:
Hi,
sorry for the delay, I have been involved in an accident with an broken leg as result, so there has been a break for some weeks.
I will in next week be able to continue, at least I hope so. There are still some outgoing issues, I will let you know asap.

fwolter commented:
Get well soon! Although, good prerequisites to do some programming.

basse04 commented:
Thank's a lot, and a very good advice I think.

basse04 commented:
Binding redesigned 20201105, by Örjan Backsell. I have now try to fulfill all the advices from @cpmeister.

BR Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,303 @@
+    @SuppressWarnings("null")
+    private void refresh() throws InterruptedException, ExecutionException, TimeoutException {
+        final @Nullable SecondGenerationBindingConstants refreshConfig;
+        refreshConfig = getConfigAs(SecondGenerationBindingConstants.class);
```
See above. Call in initialize().

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,303 @@
+                + "&dxsEntries=" + channelConfigsExtExt.get(2).dxsEntries);
+
+        // Get Gson object
+        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
```
You could make this a field to make it initialize once to save resources.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,27 @@
+	xmlns:config-description="https://openhab.org/schemas/config-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/config-description/v1.0.0 https://openhab.org/schemas/config-description-1.0.0.xsd">
+	<config-description uri="thing-type:kostalpiko1020_config">
+		<parameter name="url" type="text" required="true">
+			<label> Inverter URL</label>
+			<description>URL to inverter.</description>
+			<context>url</context>
+		</parameter>
```
The readme states the parameter is not an URL but a network address.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,27 @@
+			<description>The password for the inverter.</description>
+			<context>password</context>
+		</parameter>
+		<parameter name="refreshInterval" type="integer" min="1" unit="s" required="false">
```
This should be `required="true"`, as if the user doesn't specify it in textual configuration, it will be 0.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,431 @@
+	</channel-type>
+	<channel-type id="device-local-dc-power-pv">
+		<item-type>Number:Power</item-type>
+		<label>Dc Power PV</label>
```
Same for below.
```suggestion
		<label>DC Power PV</label>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,431 @@
+		<label>Charge Time End</label>
+		<description>Set charge time end</description>
+		<category>Energy</category>
+		<state readOnly="true"/>
```
According the readme, this and the following Channels are writeable. Should this be false?

basse04 commented:
Hi!
I have now tried to fulfill all missions you, @fwolter gave me for some weeks ago, and thank's a lot for your patience.

Regarded README.md
"If this is percent..."
It's not percent, it's text which can be sent to the inverter for configuration.

"What's the purpose..."
This is the very same configuration-parameter for the inverter, but it's possible to do the configuration in two ways.

I hope I understand your advices right, otherwise, as always, just let me know!

BR Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/README.md:
```diff
@@ -59,6 +65,77 @@ None
+| batteryType               | String          | Set battery type, Value = 1 = None, Value = 2 = PIKO Battery Li, Value = 3 = BYD B-Box HV  |   W    |
+| batteryUsageConsumption   | String          | Set battery usage consumption, Value = 100  |    W    |
+| batteryUsageStrategy      | String          | Set battery usage strategy, Value = 1 = Automatic, Value = 2 = Automatic economical  |    W    |             
+| smartBatteryControl       | Switch          | Set smart battery control, Value = False / True  |    W    |
+| smartBatteryControlText   | String          | Set smart battery control text, Value = False / True  |    W    |
```
> This is the very same configuration-parameter for the inverter, but it's possible to do the configuration in two ways.

What functionality will be missing if you remove `smartBatteryControlText` and keep only `smartBatteryControl`?

fwolter commented on bundles/org.openhab.binding.kostalinverter/README.md:
```diff
@@ -59,6 +65,77 @@ None
+|-------------------------------|---------------------------|-------------------------------------------------------|-----------------|    
+| chargeTimeEnd             | String          | Set charge time end, Value = 00:00  |    W    |
+| batteryType               | String          | Set battery type, Value = 1 = None, Value = 2 = PIKO Battery Li, Value = 3 = BYD B-Box HV  |   W    |
+| batteryUsageConsumption   | String          | Set battery usage consumption, Value = 100  |    W    |
```
> It's not percent, it's text which can be sent to the inverter for configuration.

As the readme states "Value = 100" I guess this would be a numeric value. Can you explain in short what the feature behind the parameter is?

basse04 commented:
Hi,
I have now made some changes regarded to your last issues.

SmartBatteryControl...:
I have been in contact with a user, Tobias Hieble,  ( who has a battery connected to his Piko10-20 inverter, I myself live in Sweden and it is due to taxmatters not very common with connected batteries here, and he has for some year running an earlier beta-version of this binding with SmartBatteryControlText with good result. So I decided to let this sort be the one left. But I have renamed it to SmartBatteryControl.

BatteryUsageConsumption:
Thank's to Tobias so have I edited the text.

I also found an incorrect parameter in PIKO1020.xml which now is corrected, and it's possible to see the Thing and Item's with values in MainUI.

Best regards Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -856,16 +856,9 @@
 	</channel-type>
 	<channel-type id="device-local-smart-battery-control">
-		<item-type>Switch</item-type>
+		<item-type>String</item-type>
```
Great that you received feedback from the user! It would make sense to make this a Switch if the value is true/false, although if the early adopter needs to edit his configuration.

fwolter commented on bundles/org.openhab.binding.kostalinverter/README.md:
```diff
@@ -125,10 +125,9 @@ The following Channels are writeable
 | chargeTimeEnd             | Number:Time     | Set charge time end, Value = 00:00  |    W    |
 | batteryType               | String          | Set battery type, Value = 1 = None, Value = 2 = PIKO Battery Li, Value = 3 = BYD B-Box HV  |   W    |
-| batteryUsageConsumption   | String          | Set battery usage consumption, Value = 100  |    W    |
+| batteryUsageConsumption   | String          | Set battery usage consumption level for power-consumption from battery, default value = 100 (W)  |    W    |
```
If this is watts, `String` doesn't make sense here. You could make it `Number:Power`.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/PIKO1020.xml:
```diff
@@ -3,9 +3,9 @@
+	<thing-type id="kostalinverterpiko1020">
 		<label>KOSTAL PIKO 10-20</label>
-		<description>Thing for the KOSTAL PIKO 10-20</description>
+		<description>Bindings for the KOSTAL PIKO 10-20</description>
```
Did you change this by intention? It is indeed a Thing.

basse04 commented:
I have done some changes regarded to your last issues.
About : Did you change this by intention? It is indeed a Thing. You have 20200810 asked me to change this from Binding to Thing which I then did. I have now switched back. 

About: See my comment from 7 Oct. "The binding name don't need to be included in the Thing Type ID." I have used the thing-type id from First Generation and add piko1020 (kostalinverterpiko1020), so this is the thing-type id I use for Second Generation. Do you like me to have another syntax/name for it?

Best regards,
Basse

fwolter commented:
I think it's better to use the correct notation for the newly added thing, although if it's inconsistent with the existing things.

Can you fix the DCO check?

basse04 commented:
I will make a try, will  PIKO1020  and MyPiko1020 be usable, like:
Thing kostalinverter:PIKO1020:MyPiko1020 [ url="http://'inverter-ip'", username="'username'", password="'password'", refreshInterval=60]

Regards, Basse

basse04 commented:
I have done some changes regarded to your last issues from 5 days ago.

Best regards Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -282,6 +282,7 @@ private void refresh() throws InterruptedException, ExecutionException, TimeoutE
             // Update the channels
             if (state != null) {
                 updateState(channel, state);
+                System.out.println(channel + "  " + state);
```
Leftover

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationBindingConstants.java:
```diff
@@ -29,8 +29,7 @@
     // List of all Thing Type UIDs
-    public static final ThingTypeUID SECOND_GENERATION_INVERTER = new ThingTypeUID(BINDING_ID,
-            "kostalinverterpiko1020");
+    public static final ThingTypeUID SECOND_GENERATION_INVERTER = new ThingTypeUID(BINDING_ID, "PIKO1020");
```
Thing Type IDs should be lower case.

fwolter reviewed (changes requested):
The DCO check below still fails.

basse04 commented:
"Leftover" I know it would happen soner or later, now deleted.
Changed Type ID
Done DCO, I will do my very best to remember it from now on.

//Basse

fwolter reviewed (approved):
LGTM

basse04 commented:
Found an faulty dxsId value in SecondGenerationChannelConfiguration which now is corrected. Done some changes in README.

BR Basse

basse04 commented:
Done changes regarded to Channel-type id due to multiple names.
Done some changes regarded to pattern in Channels.xml

BR Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -738,15 +738,15 @@
 	<channel-type id="statistic-own-consumption-total">
 		<item-type>Number:Energy</item-type>
-		<label>Own Consumption Total</label>
+		<label>Total own consumption</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

basse04 commented:
Now changed.
//Basse

basse04 commented:
Hi!
Done some changes mostly in SecondGenerationConfigurationHandler.
Found out that I have to change SmartHomeUnits to Units in SecondGenerationChannelConfiguration and also which I usually avoid in ThirdGenerationHandler and WebScrapeHandler.
I have to change to Build #2061 on my RPi with this version of binding to have functionality, earlier I was running #2039

BR Basse

fwolter commented:
`SmartHomeUnits` have been renamed in the main branch. Rebasing your branch could have been sufficient. There is still a conflict in your branch.

fwolter commented:
Can you give this PR a more meaningful title as it will be listed in the change log of the OH distribution without much context?

fwolter commented:
Can you fix the conflicting file? Then, this would be ready to merge.

basse04 commented:
Hi,
we have found some strange issues regarded to scraped values with battery attached to inverter I have to solve. It has no big impact to the code syntax but there will be some changes regarded to which and how the configurable values will be scraped. I will upload the new code as soon as we have test it. I need help with this though my inverter doesn't have any battery connected.
About the conflicting files, can I just use the 'Resolve conflicts' button?

BR Basse

fwolter commented:
Yes, you can use the resolve conflict button.

Please file a follow-up PR to implement the changes.

basse04 commented:
Hi!
Please have a look, I have had help to test the parts which not is possible for me due to version of inverter, and it seems so far to work as expected. It's regarded to the configuration possibilities of the binding-part for Piko10-20.

Best Regards Basse

fwolter commented:
What did you change since the last commit? Please don't squash commits in the PR, as the reviewer can't see the changes and need to review the whole code again.

You need to fix the conflicting file before this can get merged.

basse04 commented:
Thank's for the info.

SecondGenerationBindingConstants.java
Added and deleted channels.

SecondGenerationChannelConfiguration.java
Added channels.

SecondGenerationHandler.java
added ArrayList, line 68
added if(... , line 89
redesigned  line 90-146
added posts for ArrayList,  line 156
redesigned refresh() part, line 202-292
added if(... , line 294
redsigned line 297-355
added method stringToSeconds(... , line 393-398

README.md
Changes done due to above.

Please just let me know if you need more info!!

How do I easiest/best fix the conflict?
I can see the differences in code but can I just change it and do another commit or will this lead to mismatch?

Best Regards Basse

fwolter commented:
You could change it in code manually or use git rebase and resolve the conflicts, then. Whatever is easier for you.

basse04 commented:
Redesigned due to  conflicting files.

fwolter commented:
Your branch is still conflicting.

basse04 commented:
Redesigned due to conflicting files. But Please help me, I can't see reason for described Conflicting files. 
BR Basse

fwolter commented:
Seems nothing fancy. You can click on the "resolve conflicts" button to show and fix the conflicts.

basse04 commented:
A question, when I have pushed the Resolve conflicts button and redesigned the conflicts, pushed the Mark as resolved button. Should I then push the Commit Merge button , there is warnings, that is  why I ask?

fwolter commented:
Seems like this is necessary, as the conflicting file is still shown up.

basse04 commented:
Done as described above.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+        this.httpClient = httpClient;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String url = inverterConfig.url;
```
The field `inverterConfig` can be null at this point. To mitigate this, you can store the field to a local variable before checking it:
```

Future<?> localInverterConfig = inverterConfig;
if (localInverterConfig != null) {
    localInverterConfig .cancel(true);
}
```

Then, you can remove the supression.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationConfigurationHandler.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    // Method extractSessionId extracts sessionId from JsonObject
+    static String extractSessionId(JsonObject extractJsonObjectSessionId) throws TimeoutException {
```
This exception is never thrown and can therefore be removed:
```suggestion
    static String extractSessionId(JsonObject extractJsonObjectSessionId) {
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationConfigurationHandler.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    // Method extractCode extracts code from JsonObject
+    static String extractCode(JsonObject extractJsonObjectCode) throws TimeoutException {
```
See above.
```suggestion
    static String extractCode(JsonObject extractJsonObjectCode) {
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+            } catch (InterruptedException e) {
+                logger.debug("Communication with inverter interrupted, exception {}", e.getMessage());
+            } catch (ExecutionException e) {
+                logger.debug("Communication with inverter failed, exception {}", e.getMessage());
+            } catch (TimeoutException e) {
+                logger.debug("Communication with inverter timed out, exception {}", e.getMessage());
```
You might want to set the Thing offline when the connection fails.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+                refresh();
+                updateStatus(ThingStatus.ONLINE);
+            } catch (RuntimeException scheduleWithFixedDelayException) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        scheduleWithFixedDelayException.getClass().getName() + ":"
+                                + scheduleWithFixedDelayException.getMessage());
+                logger.debug("Error when connecting to inverter, {}: {}", getThing().getUID(),
+                        scheduleWithFixedDelayException.getMessage());
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+        this.inverterConfig = inverterConfig;
+
+        // Initialize Gson object gson
+        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        this.gson = gson;
```
You could initialize this in the constructor to get rid of the compiler warning about null values. You don't see them currently as you supressed them in `refresh()`.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+    @SuppressWarnings({ "null", "unchecked" })
+    private void refresh() throws InterruptedException, ExecutionException, TimeoutException {
+        // dxsEntries part
+        String dxsEntriesCall = inverterConfig.url.toString() + "/api/dxs.json?dxsEntries="
```
This generates a compiler warning, which is currently suppressed. You could store `inverterConfig` to a local variable and do a null check on that to get rid of the warning.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+            channelPosts[channelPostsCounter] = dxsentries.getName();
+            channelPostsCounter++;
+        }
+        List<String> channelPostsTemp = new ArrayList<String>(Arrays.asList(channelPosts));
```
Syntactical sugar. Same for the others.
```suggestion
        List<String> channelPostsTemp = List.of(channelPosts);
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,410 @@
+        List<String> channelPostsTemp = new ArrayList<String>(Arrays.asList(channelPosts));
+
+        // dxsEntriesExt part
+        String dxsEntriesCallExt = inverterConfig.url.toString() + "/api/dxs.json?dxsEntries="
```
`toString()` is called implicitly and could therefore be removed. Please check all.
```suggestion
        String dxsEntriesCallExt = inverterConfig.url + "/api/dxs.json?dxsEntries="
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,31 @@
+		<parameter name="url" type="text" required="true">
+			<label> IP Address</label>
+			<description>IP address of the inverter.</description>
+			<context>network-address</context>
```
If I see correctly this is a URL and not a hostname.
```suggestion
			<context>url</context>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,31 @@
+			<description>Refresh Interval in seconds.</description>
+			<default>60</default>
+		</parameter>
+		<parameter name="invertertype" type="text" required="true">
+			<label>Inverter Typel</label>
+			<description>Type of inverter, with/without battery.</description>
+		</parameter>
```
Why not make this a boolean `hasBattery`?

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,438 @@
+	</channel-type>
+	<channel-type id="device-local-akt-home-consumption-solar">
+		<item-type>Number:Power</item-type>
+		<label>Actual Home Consumption Solar</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,438 @@
+	</channel-type>
+	<channel-type id="device-local-max-depth-of-discharge">
+		<item-type>Number:Dimensionless</item-type>
+		<label>Max Depth Of Discharge</label>
```
```suggestion
		<label>Max Depth of Discharge</label>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/KostalInverterFactory.java:
```diff
@@ -81,7 +85,9 @@ public KostalInverterFactory(@Reference final HttpClientFactory httpClientFactor
     @Override
     public boolean supportsThingType(ThingTypeUID thingTypeUID) {
-        return thingTypeUID.equals(FIRST_GENERATION_INVERTER)
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
```
Seems like there had been something wrong during merge. This produces a compiler error.

fwolter reviewed (changes requested):
Great that the conflicts are solved!

It's better to not use `git merge` in pull requests, because it could garble your Git history. You could use `git rebase` instead.

basse04 commented:
Done some changes regarded to advices. 
I could not handle Future...., so I done the changes as the other part of binding.
Try to do the Labels as small as possible without taken away info.
Think I found the fault in KostalInverterFactory.java , was somehow duplicated lines.

I use to use Commit and Push from within Eclipse, do you think it's better to use git rebase?

BR Basse

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,406 @@
+    private List<String> channelPostsTempExtExt = new ArrayList<>();
+    private List<String> channelPostsTempAll = new ArrayList<>();
+
+    private @NonNullByDefault({}) SecondGenerationInverterConfig inverterConfig;
```
You could initialize the field with an empty config. Then, you can remove the null annotation.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/kostalinverter/internal/secondgeneration/SecondGenerationHandler.java:
```diff
@@ -0,0 +1,406 @@
+    private List<String> channelPostsTempAll = new ArrayList<>();
+
+    private @NonNullByDefault({}) SecondGenerationInverterConfig inverterConfig;
+    private @NonNullByDefault({}) Gson gson;
```
You could initialize the field to remove the null annoation.
```suggestion
    private Gson gson = new GsonBuilder().setPrettyPrinting().create();
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/config/SecondGeneration.xml:
```diff
@@ -0,0 +1,31 @@
+	xsi:schemaLocation="https://openhab.org/schemas/config-description/v1.0.0 https://openhab.org/schemas/config-description-1.0.0.xsd">
+	<config-description uri="thing-type:kostalpiko1020_config">
+		<parameter name="url" type="text" required="true">
+			<label> IP Address</label>
```
```suggestion
			<label>IP Address</label>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,438 @@
+	</channel-type>
+	<channel-type id="device-local-akt-home-consumption-solar">
+		<item-type>Number:Power</item-type>
+		<label>Akt Home Consumption Solar</label>
```
What do you mean by "Akt"? Would "Home Consumption PV" suffice? Same for below.

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,438 @@
+	</channel-type>
+	<channel-type id="device-local-battery-charge-time-from-set">
+		<item-type>String</item-type>
+		<label>Battery Charge Time From Set</label>
```
```suggestion
		<label>Battery Charge Time from Set</label>
```

fwolter commented on bundles/org.openhab.binding.kostalinverter/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -484,4 +484,438 @@
+	</channel-type>
+	<channel-type id="device-local-battery-charge-time-to-set">
+		<item-type>String</item-type>
+		<label>Battery Charge Time To Set</label>
```
```suggestion
		<label>Battery Charge Time to Set</label>
```

fwolter commented:
> I use to use Commit and Push from within Eclipse, do you think it's better to use git rebase?

That's totally fine. Only merge should be avoided in pull requests.

basse04 commented:
Thank's for your advice, I will then continue to do as before, using the Eclipse way.
Please, have a look at the latest changes.

fwolter reviewed (approved):
Your latest changes look good. Let's get this merged finally!

basse04 commented:
That's Great news.
Yesterday  where I noticed about a bug regarded Kostalinverter vs Zwave, I have seen this topic before but I thought it was just regarded to Third Generation. But it is affecting the entire Kostalinverter binding.  Please, have a look at #7090.
I think I have found a solution, and my question is, how do you prefer me to handle this issue.
Upload the changes here or in #7090?

BR Basse

fwolter commented:
As this is already merged, it cannot receive any further changes. It's good to go if you open a new pull request for fixing this issue.
